// wipe-node-cache/lib/index.js
"use strict";

function waveCallback_default() {
  return true;
}

function removeFromCache_nodejs(moduleName) {
  delete require.cache[moduleName];
}

function assignParents(modules) {
  var result = {};
  Object.keys(modules).forEach(function (moduleName) {
    var parent = modules[moduleName];
    var line = parent.children;
    line.forEach(function (_ref) {
      var childName = _ref.id;

      result[childName] = result[childName] || { parents: [] };
      result[childName].parents.push(moduleName);
    });
  });
  return result;
}

function burn(cache, wipeList, lookup, callback) {
  var removeFromCache = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : removeFromCache_nodejs;

  // Secondary wave
  while (wipeList.length) {
    var removeList = wipeList;
    wipeList = [];

    removeList.forEach(function (moduleName) {
      if (callback(moduleName)) {
        if (lookup[moduleName]) {
          wipeList.push.apply(wipeList, lookup[moduleName].parents);
          delete lookup[moduleName];
        }

        removeFromCache(moduleName);
      }
    });
  }
}

function purge(cache, wipeList, callback, removeFromCache) {
  burn(cache, wipeList, assignParents(cache), callback, removeFromCache);
}

function reverseString(str) {
  var result = "";
  for (var i = str.length - 1; i >= 0; i--) {
    result += str[i];
  }
  return result;
}

function buildIndexForward(cache) {
  return Object.keys(cache);
}

function getCache() {
  return require.cache;
}

/**
 * Wipes node.js module cache.
 * First it look for modules to wipe, and wipe them.
 * Second it looks for users of that modules and wipe them to. Repeat.
 * Use waveCallback to control secondary wave.
 * @param {Object} stubs Any objects, which will just be passed as first parameter to resolver.
 * @param {Function} resolver function(stubs, moduleName) which shall return true, if module must be wiped out.
 * @param {Function} [waveCallback] function(moduleName) which shall return false, if parent module must not be wiped.
 */
function wipeCache(stubs, resolver, waveCallback) {
  var removeFromCache = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : removeFromCache_nodejs;

  waveCallback = waveCallback || waveCallback_default;
  var cache = require.cache;

  wipeMap(cache, function (cache, callback) {
    return cache.forEach(function (moduleName) {
      return resolver(stubs, moduleName) && callback(moduleName);
    });
  }, waveCallback, removeFromCache);
}

function wipeMap(cache, callback, waveCallback, removeFromCache) {
  var wipeList = [];
  callback(buildIndexForward(cache), function (name) {
    removeFromCache(name);
    wipeList.push(name);
  });
  return purge(cache, wipeList, waveCallback);
}

exports.buildIndexForward = buildIndexForward;
exports.getCache = getCache;
exports.removeFromCache = removeFromCache_nodejs;
exports.purge = purge;
exports.burn = burn;

exports.wipeCache = wipeCache;
exports.wipeMap = wipeMap;
